### PR TITLE
[Debugger] Move the Download-via-SourceLink prompt into the NoSourceV…

### PR DESCRIPTION
**Pull request by jstedfast**
…iew page

Instead of annoying the user with a dialog prompt and then,
when the user clicks Cancel, showing them the NoSourceView
document - just add a "Download via SourceLink" button to
the NoSourceView document tab allowing the user to download
the source there.

This simplifies the process and makes it much less annoying.

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/966732/

### CODE REVIEW

**Comment by jstedfast** [comment]
download-via-sourcelink

**Comment by nosami** [comment]
"Step into External Code" is a feature that needs to be enabled manually first so not sure why it would be annoying?

And now, there is no easy to find option to automatically download? I think if you wanted to the UI to work this way then there should be a 3rd button to set the "always download" property.

File: main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/NoSourceView.cs [hunk 1]
```diff
@@ -23,49 +23,71 @@
...
+
 				var actionsBox = new HBox ();
+
 				var buttonBrowseAndFind = new Button (GettextCatalog.GetString ("Browse and find {0}", fileName));
 				buttonBrowseAndFind.Clicked += OpenFindSourceFileDialog;
 				actionsBox.PackStart (buttonBrowseAndFind);
+
+				if (frame.SourceLocation?.SourceLink != null && options.AutomaticSourceLinkDownload == AutomaticSourceDownload.Ask) {
+					var buttonDownload = new Button (GettextCatalog.GetString ("Download via Source Link"));
+					buttonDownload.Clicked += DownloadSourceLink;
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/NoSourceView.cs]: Handler should be unsubscribed.

**Comment by jstedfast** [comment]
@vancura can you review this UI change and make suggestions?

**Comment by nosami** [comment]
> @vancura can you review this UI change and make suggestions?

@hbons probably should take a look too seeing as he was the person involved initially with the design.

**Comment by sgmunn** [comment]
@vancura the original bug suggested to add an option to automatically do this, which was on the original dialog. should we add that here for completeness?

**Comment by vancura** [comment]
@sgmunn Yes please

**Comment by jstedfast** [comment]
Updated screenshot:

sourcelink-checkbox

**Comment by vancura** [comment]
LGTM. @hbons?

**Comment by jstedfast** [comment]
The Leak Tests failure seems to be unrelated (failure to install graphviz)

**Comment by hbons** [comment]
@jstedfast I like your idea. I just made a few tweaks to the presentation:
image
image

thoughts?

**Comment by jstedfast** [comment]
@hbons I think I like your design better than the existing UI.

**Comment by jstedfast** [comment]
Updated UI based on @hbons comment:

sourcelink

File: main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/NoSourceView.cs [hunk 2]
```diff
@@ -23,91 +23,173 @@
...
+						MarginBottom = 20,
+						MarginTop = 10,
+						Spacing = 10,
+					};
+
+					var label = new Label {
+						Markup = GettextCatalog.GetString ("External source code is available. Would you like to download {0} and view it?", $"<a href=\"clicked\">{fileName}</a>"),
+						Name = "SourceLinkLabel"
+					};
+					label.LinkClicked += DownloadSourceLink;
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger/NoSourceView.cs]: Event handlers need to be unsubscribed.

**Pull request merged**